### PR TITLE
Add SetItem function for __setitem__ of Variable

### PR DESCRIPTION
**Pull request by wkentaro**
Inverse operation of GetItem function: https://github.com/pfnet/chainer/pull/1154

### CODE REVIEW

**Comment by beam2d** [comment]
Thank you for the PR, but I don't think this setitem works correctly with backprop. For example, consider a statement `x[i] = v` for Variables x and v with some slices i. The resulting x has an updated array but the graph rooted with x is not modified (I mean, x does not depend on v on the graph even after the assignment). Thus, when we try to compute a gradient of a loss function `f(x)`, the error is not propagated to v, although the loss actually depends on v.

One possible idea to resolve this problem is to modify the graph by replacing the node x by its copy and inserting the SetItem node to the graph as a new creator of the original x.

Another concern is that some Functions (e.g. sigmoid, tanh, and cuDNN relu) keep its output array for backprop and assume that the array is not modified, and so applying setitem to the output of these Functions break the correctness of backprop. It would be very difficult to detect such a misuse.
